pkg/operator/ceph/csi/cluster_config.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	ccJson, err := json.Marshal(cc)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal csi cluster config. %+v", err)
+	}
+	return string(ccJson), nil
+}
+
+func parseCsiClusterConfig(c string) (csiClusterConfig, error) {
+	var cc csiClusterConfig
+	err := json.Unmarshal([]byte(c), &cc)
+	if err != nil {
+		return cc, fmt.Errorf("failed to parse csi cluster config. %+v", err)
+	}
+	return cc, nil
+}
+
+func formatCsiClusterConfig(cc csiClusterConfig) (string, error) {
+	ccJson, err := json.Marshal(cc)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal csi cluster config. %+v", err)
+	}
+	return string(ccJson), nil
+}
+
+func monEndpoints(mons map[string]*cephconfig.MonInfo) []string {
+	endpoints := make([]string, 0)
+	for _, m := range mons {
+		endpoints = append(endpoints, m.Endpoint)
+	}
+	return endpoints
+}
+
+// UpdateCsiClusterConfig returns a json-formatted string containing
+// the cluster-to-mon mapping required to configure ceph csi.
+func UpdateCsiClusterConfig(
+	curr, clusterKey string, mons map[string]*cephconfig.MonInfo) (string, error) {
+
+	var (
+		cc     csiClusterConfig
+		centry csiClusterConfigEntry
+		found  bool
+	)
+	cc, err := parseCsiClusterConfig(curr)
+	if err != nil {
+		return "", err
+	}
+
+	for i, centry := range cc {
+		if centry.ClusterID == clusterKey {
+			centry.Monitors = monEndpoints(mons)
+			found = true
+			cc[i] = centry
+			break
+		}
+	}
+	if !found {
+		centry.ClusterID = clusterKey
+		centry.Monitors = monEndpoints(mons)
+		cc = append(cc, centry)
+	}
+	return formatCsiClusterConfig(cc)
+}
+
+// CreateCsiConfigMap creates an empty config map that will be later used
+// to provide cluster configuration to ceph-csi.
+func CreateCsiConfigMap(namespace string, clientset kubernetes.Interface) error {
+	configMap := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ConfigName,
+			Namespace: namespace,
+		},
+	}
+	configMap.Data = map[string]string{
+		ConfigKey: "[]",
+	}
+
+	if _, err := clientset.CoreV1().ConfigMaps(namespace).Create(configMap); err != nil {
+		if !k8serrors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+	return nil
+}
+
+// SaveClusterConfig updates the config map used to provide ceph-csi with
+// basic cluster configuration. The clusterNamespace and clusterInfo are
+// used to determine what "cluster" in the config map will be updated and
+// and the clusterNamespace value is epxected to match the clusterID
+// value that is provided to ceph-csi uses in the storage class.
+// The locker l is typically a mutex and is used to prevent the config
+// map from being updated for multiple clusters simultaneously.
+func SaveClusterConfig(
+	clientset kubernetes.Interface, clusterNamespace string,
+	clusterInfo *cephconfig.ClusterInfo, l sync.Locker) error {
+
+	if !CSIEnabled() {
+		return nil
+	}
+	l.Lock()
+	defer l.Unlock()
+	// csi is deployed into the same namespace as the operator
+	csiNamespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
```

Comment:
what if `PodNamespaceEnvVar` is empty?

Reply:
We can check if it is empty and error out.

Reply:
Yes please.